packages/core/src/services/agents/run.ts

```diff
@@ -1,166 +1,105 @@
+  const chainStartTime = Date.now()
 
-  let responseResolve: (value: ChainResponse<StreamType>) => void
-
-  const response = new Promise<ChainResponse<StreamType>>((resolve) => {
-    responseResolve = resolve
+  const chainStreamManager = new ChainStreamManager({
+    errorableUuid,
+    messages: pausedMessages,
+    tokenUsage: pausedTokenUsage,
   })
 
-  let conversation: Conversation
   let stepCount = 0
-  let previousResponse = _previousResponse
-  let extraMessages = _extraMessages
-
-  const chainResult = await runChain({
-    workspace,
-    chain,
-    promptlVersion,
-    providersMap,
-    source,
-    errorableType: errorableType as ErrorableEntity,
-    configOverrides,
-    persistErrors: persistErrors as true,
-    generateUUID: () => errorableUuid,
-    removeSchema: true, // Removes the schema configuration for the AI generation, as it is reserved for the agent's Return function
-    previousResponse,
-    extraMessages,
-    ...other,
-  })
-
-  const chainStartTime = Date.now()
-  const stream = new ReadableStream<ChainEvent>({
-    start(controller) {
-      const chainEventsReader = chainResult.stream.getReader()
-
-      const readNextChainEvent = () => {
-        chainEventsReader.read().then(({ done, value }) => {
-          if (value) {
-            if (value.data.type === ChainEventTypes.Error) {
-              controller.enqueue(value)
-              return responseResolve(
-                Result.error(value.data.error as ChainError<RunErrorCodes>),
-              )
-            }
-
-            if (isChainComplete(value)) {
-              conversation = {
-                config: value.data.config,
-                messages: value.data.response.providerLog?.messages ?? [],
-              }
 
-              // Pause this process if the chain stopped due to tool calls
-              if (value.data.finishReason === 'tool-calls') {
-                controller.enqueue(value)
-                controller.close()
-                return responseResolve(
-                  Result.ok(
-                    value.data.response as ChainStepResponse<StreamType>,
-                  ),
-                )
-              }
-            } else {
-              // Forward all events that are not the ChainComplete event
-              controller.enqueue(value)
-
-              if (value.data.type === ChainEventTypes.StepComplete) {
-                stepCount++
-                extraMessages = undefined // extraMessages has been used by the Chain, so we no longer need it for the agent workflow
-                previousResponse = value.data
-                  .response as ChainStepResponse<StreamType>
-              }
-            }
-          }
-
-          if (!done) return readNextChainEvent()
-
-          // Chain has been completed. Remove the cached chain if there was any.
-          deleteCachedChain({ workspace, documentLogUuid: errorableUuid })
+  const streamResult = chainStreamManager.start(async () => {
```

Comment:
Here you are throwing errors inside `.start`. I would always return a Result and do the `.unwrap()` here. Let's not put handled errors inside services. 

The less try/catch we have the better